src/lib/utilities-for-location.ts

```diff
@@ -10,46 +18,31 @@ import { GeoInfo } from './income-info';
  *
  * Returns a map of utility IDs to utility info, suitable for the response
  * from /api/v1/utilities.
- *
- * TODO this is very not scalable to nationwide coverage!
  */
-export function getUtilitiesForLocation(
+export async function getUtilitiesForLocation(
+  db: Database,
   location: GeoInfo,
   includeBeta: boolean,
-): {
+): Promise<{
   [id: string]: Authority;
-} {
+}> {
   const stateUtilities = AUTHORITIES_BY_STATE[location.state_id]?.utility;
 
   if (!stateUtilities || !isStateIncluded(location.state_id, includeBeta)) {
     return {};
   }
 
-  let ids: string[];
-
-  // If we know details about where the state's utilities operate, we can
-  // encode that here.
-  if (location.state_id === 'RI') {
-    // Source: https://catalog.data.gov/dataset/u-s-electric-utility-companies-and-rates-look-up-by-zipcode-2021
-    // According to that dataset, 02839 is not served by Pascoag, but in a meeting
-    // with them they mentioned it, so it's included here.
-    switch (location.zip) {
-      case '02807':
-        ids = ['ri-block-island-power-company'];
-        break;
-      case '02814':
-      case '02830':
-      case '02839':
-      case '02859':
-        ids = ['ri-rhode-island-energy', 'ri-pascoag-utility-district'];
-        break;
-      default:
-        ids = ['ri-rhode-island-energy'];
-    }
-  } else {
-    // Fall back to returning all the state's utilities, regardless of location.
-    return stateUtilities;
-  }
+  // Put the predominant utility first
+  const rows = await db.all<ZipToUtility[]>(
+    'SELECT * FROM zip_to_utility WHERE zip = ? ORDER BY predominant DESC',
+    location.zip,
```

Comment:
If there are no rows found, I think we should return all of the state's utilities in authorities.json as fallback – right now this will just return empty.

Definitely not part of this PR, but what happens if the dataset is wrong and the user is not offered a list that includes their utility? Do they have any way of letting us know/requesting a full state list instead?

It might also be good to do a quick coverage test to make sure this has good coverage across our zips. It looks good overall, but I think that's something to ensure before we go ahead.

Reply:
I wondered about how to handle the possibility of not showing a user's actual utility in my notes doc. Regardless of zip->utility mapping errors, we know there are utilities that even this dataset doesn't contain (i.e. there are several Vermont munis that the Vermont PUC lists that aren't in this dataset, as noted in the notes doc).

I think I would be satisfied, at least in the short term, with including an "Other" item in the frontend's utility selector, but not having a way for them to tell us what their utility is, since that would be a nontrivial design change. And maybe we can monitor for "Other" responses and see what zips they're in, and investigate from there?

We can include checking the utility options in the per-state QA process. I'll add that to the launch checklist.